packages/flutter/lib/src/gestures/resampler.dart

```diff
@@ -0,0 +1,428 @@
+      // or equal to `sampleTime`.
+      if (event.timeStamp <= sampleTime || _last == null) {
+        _last = event;
+        _next = event;
+        continue;
+      }
+
+      // Update only `next` pointer event if time stamp is more recent than
+      // `sampleTime` and next event is not already more recent.
+      final Duration nextTimeStamp = _next?.timeStamp ?? const Duration();
+      if (nextTimeStamp < sampleTime) {
+        _next = event;
+        break;
+      }
+    }
+  }
+
+  List<PointerEvent> _dequeueAndSamplePointerEventsUntil(Duration sampleTime) {
+    final List<PointerEvent> dequeuedEvents = <PointerEvent>[];
+
+    while (_queuedEvents.isNotEmpty) {
+      final PointerEvent event = _queuedEvents.first;
+
+      // Potentially stop dequeuing events if more recent than `sampleTime`.
+      if (event.timeStamp > sampleTime) {
+        // Stop if event is not up or removed. Otherwise, continue to
+        // allow early processing of up and remove events as this improves
+        // resampling of these events, which is important for fling
+        // animations.
+        if (event is! PointerUpEvent && event is! PointerRemovedEvent) {
+          break;
+        }
+
+        // When this line is reached, the following two invariants hold:
+        // (1) `event.timeStamp > sampleTime`
+        // (2) `_next` has the smallest time stamp that's no less than
+        //     `sampleTime`
+        //
+        // Therefore, event must satisfy `event.timeStamp >= _next.timeStamp`.
+        //
+        // Those events with the minimum `event.timeStamp == _next.timeStamp`
+        // time stamp are processed early for smoother fling. For events with
+        // `event.timeStamp > _next.timeStamp`, the following lines break the
+        // while loop to stop the early processing.
+        //
+        // Specifically, when `sampleTime < _next.timeStamp`, there must be
+        // at least one event with `_next.timeStamp == event.timeStamp`
+        // and that event is `_next` itself, and it will be processed early.
+        //
+        // When `sampleTime == _next.timeStamp`, all events with
+        // `event.timeStamp > sampleTime` must also have
+        // `event.timeStamp > _next.timeStamp` so no events will be processed
+        // early.
+        //
+        // When the input frequency is no greater than the sampling
+        // frequency, this early processing should guarantee that `up` and
+        // `remove` events are always re-sampled.
+        final Duration nextTimeStamp = _next?.timeStamp ?? const Duration();
+        assert(event.timeStamp >= nextTimeStamp);
+        if (event.timeStamp > nextTimeStamp) {
+          break;
+        }
+      }
+
+      final bool wasTracked = _isTracked;
+      final bool wasDown = _isDown;
+
+      // Update `_Pointer` state.
+      if (event is PointerDownEvent || event is PointerMoveEvent) {
+        _isDown = true;
```

Comment:
Is this `_isDown` a duplicate of `event.down`?

Reply:
Yes, I cleaned up this code in latest version. Thanks for the hint.